Source/Tools/LinuxEmulation/LinuxSyscalls/SyscallsSMCTracking.cpp

```diff
@@ -153,6 +153,177 @@ FEXCore::HLE::AOTIRCacheEntryLookupResult SyscallHandler::LookupAOTIRCacheEntry(
+    //       us to be more optimal by using GuardSignalDeferringSection instead
+    auto lk = FEXCore::GuardSignalDeferringSectionWithFallback(VMATracking.Mutex, Thread);
+
+    bool Map32Bit = !Is64Bit || (flags & FEX::HLE::X86_64_MAP_32BIT);
+    if (Map32Bit) {
+      Result = (uint64_t)Get32BitAllocator()->Mmap((void*)addr, length, prot, flags, fd, offset);
+      if (FEX::HLE::HasSyscallError(Result)) {
+        return reinterpret_cast<void*>(Result);
+      }
+      LOGMAN_THROW_A_FMT(Is64Bit || (Result >> 32) == 0 || (Result >> 32) == 0xFFFFFFFF, "values must fit to 32 bits");
+    } else {
+      Result = reinterpret_cast<uint64_t>(::mmap(reinterpret_cast<void*>(addr), length, prot, flags, fd, offset));
+      if (Result == ~0ULL) {
+        return reinterpret_cast<void*>(-errno);
+      }
+    }
+
+    FEX::HLE::_SyscallHandler->TrackMmap(Thread, Result, length, prot, flags, fd, offset);
+  }
+
+  FEX::HLE::_SyscallHandler->InvalidateCodeRangeIfNecessary(Thread, Result, Size);
+  return reinterpret_cast<void*>(Result);
+}
+
+uint64_t SyscallHandler::GuestMunmap(bool Is64Bit, FEXCore::Core::InternalThreadState* Thread, void* addr, uint64_t length) {
+  LOGMAN_THROW_A_FMT(Is64Bit || (reinterpret_cast<uintptr_t>(addr) >> 32) == 0, "values must fit to 32 bits: {}", fmt::ptr(addr));
+  LOGMAN_THROW_A_FMT(Is64Bit || (length >> 32) == 0, "values must fit to 32 bits");
+
+  uint64_t Result {};
+  uint64_t Size = FEXCore::AlignUp(length, FEXCore::Utils::FEX_PAGE_SIZE);
+
+  {
+    // Frontend calls this with nullptr Thread during initialization.
+    // This is why `GuardSignalDeferringSectionWithFallback` is used here.
+    // To be more optimal the frontend should provide this code with a valid Thread object earlier.
+    auto lk = FEXCore::GuardSignalDeferringSectionWithFallback(VMATracking.Mutex, Thread);
+
+    if (reinterpret_cast<uintptr_t>(addr) < 0x1'0000'0000ULL) {
+      Result = FEX::HLE::_SyscallHandler->Get32BitAllocator()->Munmap(addr, length);
+      if (FEX::HLE::HasSyscallError(Result)) {
+        return Result;
+      }
+    } else {
+      Result = ::munmap(addr, length);
+      if (Result == -1) {
+        return -errno;
+      }
+    }
+    FEX::HLE::_SyscallHandler->TrackMunmap(Thread, addr, length);
+  }
+  FEX::HLE::_SyscallHandler->InvalidateCodeRangeIfNecessary(Thread, reinterpret_cast<uint64_t>(addr), Size);
+
+  return Result;
+}
+
+uint64_t SyscallHandler::GuestMremap(bool Is64Bit, FEXCore::Core::InternalThreadState* Thread, void* old_address, size_t old_size,
+                                     size_t new_size, int flags, void* new_address) {
+  uint64_t Result {};
+
+  {
+    auto lk = FEXCore::GuardSignalDeferringSection(FEX::HLE::_SyscallHandler->VMATracking.Mutex, Thread);
+    if (Is64Bit) {
+      Result = reinterpret_cast<uint64_t>(::mremap(old_address, old_size, new_size, flags, new_address));
+      if (Result == -1) {
+        return Result;
```

Comment:
This one is a little weird, it should return return `-errno` here so that it always returns combined errno result.
Then on the 64-bit syscall handler it can just return the result directly.
Currently this code is a weird edge of 64-bit returns result plus errno, where 32-bit returns single result.